### PR TITLE
[forge] Add python forge wrapper to replace run_forge

### PR DESCRIPTION
**Pull request by perryjrandall**
Run forge is hard to read / test, since we have a wrapper make it maintainable.

Longer term we want to move most of this functionality to the forge cli itself, but in the meantime we shouldnt have to compile to handle certain tasks around forge execution.

Test Plan:

Running locally / forge on PR

Unittests added pass as well as #e2e

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2992)

### CODE REVIEW

File: testsuite/run_forge.sh [hunk 1]
```diff
@@ -9,6 +9,18 @@
 # if the necessary environment variables are set.
 #
 
+# Calculate a shard for cutting over cleanly
+SHARD_KEY="${GITHUB_USER:-$USER}"
+SHARD="$(echo $SHARD_KEY | PYTHONHASHSEED=1 python3 -c "print(abs(hash(input())) % 100)")"
+MAX_SHARD="${MAX_SHARD:-0}"
```
Comment (rustielin) [inline testsuite/run_forge.sh:15]: where are you defining `MAX_SHARD`? do you mean to have it default to 0 at first?

File: testsuite/forge.py [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+    for line in result.unwrap().decode().splitlines():
+        if line.startswith("AWS_"):
+            key, val = line.split("=", 1)
+            os.environ[key] = val
+    return get_aws_account_num(shell)
+
+
+@main.command()
+# for calculating regression in local mode
+@envoption("LOCAL_P99_LATENCY_MS_THRESHOLD", "60000")
```
Comment (rustielin) [inline testsuite/forge.py:447]: dang this is pretty cool

File: testsuite/forge.py [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+try:
+    import click
+except ImportError:
+    install_dependency("click")
+    import click
+
+try:
+    import psutil
+except ImportError:
+    install_dependency("psutil")
```
Comment (rustielin) [inline testsuite/forge.py:82]: we'll need to ensure these are installed in CI as well

**Comment by zekun000** [comment]
not a rust cli?

File: testsuite/forge.py [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+    except Exception as e:
+        return "Forge report malformed: {}\n{}".format(e, '\n'.join(report_lines))
+    if not report_text:
+        return "Forge report text empty. See test runner output."
+    else:
+        return report_text
+
+
+def get_validator_logs_link() -> str:
+    return "https://banana"
```
Comment (sitalkedia) [inline testsuite/forge.py:283]: needs to be updated..

File: testsuite/forge.py [hunk 5]
```diff
@@ -0,0 +1,559 @@
...
+
+
+def get_dashboard_link() -> str:
```
Comment (sitalkedia) [inline testsuite/forge.py:286]: same as above

File: testsuite/run_forge.sh (hunk 1 above)
Comment (perryjrandall) [reply testsuite/run_forge.sh:15]: This allows it to be passed via env var or default to 0 which is off

File: testsuite/forge.py (hunk 2 above)
Comment (perryjrandall) [reply testsuite/forge.py:447]: Sheeee

File: testsuite/forge.py (hunk 3 above)
Comment (perryjrandall) [reply testsuite/forge.py:82]: How can I do that? This actually should be able to do it inline

File: testsuite/forge.py (hunk 4 above)
Comment (perryjrandall) [reply testsuite/forge.py:283]: Yeah this is a draft pr

**Comment by perryjrandall** [comment]
> not a rust cli?

One day I will be as good at rust as da kun

**Comment by perryjrandall** [comment]
> @perryjrandall looks like failure in auth :https://github.com/aptos-labs/aptos-core/runs/7869128583?check_suite_focus=true

Woops the code for checking token auth was overly protective for CI

File: .github/workflows/build-images.yaml [hunk 6]
```diff
@@ -32,6 +32,7 @@ on: # build on main branch OR when a PR is labeled with `CICD:build-images`
       - canary
       - devnet
       - testnet
+      - chaotic
```
Comment (rustielin) [inline .github/workflows/build-images.yaml:35]: remember to remove this

File: testsuite/forge.py [hunk 7]
```diff
@@ -0,0 +1,815 @@
...
+        (result.start_time, result.end_time),
+    )
+    humio_logs_link = get_humio_logs_link(context.forge_namespace)
+
+
+    return textwrap.dedent(
+        f"""
+        =====START FORGE COMMENT=====
+        ```
+        ```
```
Comment (rustielin) [inline testsuite/forge.py:500]: move these triple ticks down?

File: testsuite/run_forge.sh [hunk 8]
```diff
@@ -9,18 +9,32 @@
...
+# REMOVE THIS
+WRAPPER_KILLSWITCH=false
 
 # output files
 FORGE_OUTPUT=${FORGE_OUTPUT:-$(mktemp)}
 FORGE_REPORT=${FORGE_REPORT:-$(mktemp)}
 FORGE_PRE_COMMENT=${FORGE_PRE_COMMENT:-$(mktemp)}
 FORGE_COMMENT=${FORGE_COMMENT:-$(mktemp)}
 
+if [[ $WRAPPER_KILLSWITCH = true ]]; then
```
Comment (rustielin) [inline testsuite/run_forge.sh:23]: If you expect these to be env vars, you'll need to add it to `run-forge.yaml` as an env that's populated from GH secret

File: .github/workflows/run-forge.yaml [hunk 9]
```diff
@@ -65,6 +65,13 @@ jobs:
           ref: ${{ inputs.GIT_SHA }}
           # get the last 10 commits if IMAGE_TAG is not specified
           fetch-depth: env.IMAGE_TAG != null && 0 || 10
+      - uses: actions/setup-python@v4
+        if: env.FORGE_ENABLED == 'true'
+        with:
+          python-version: '3.9'
```
Comment (banool) [inline .github/workflows/run-forge.yaml:71]: Let's not do this, this step reads from `.python_version` (which is also 3.9). Good to keep a uniform version.

File: testsuite/forge.py [hunk 10]
```diff
@@ -0,0 +1,815 @@
...
+    def run(self, command: Sequence[str], stream_output: bool = False) -> RunResult:
+        return RunResult(0, b'output')
+
+
+def install_dependency(dependency: str) -> None:
+    print(f"{dependency} is not currently installed")
+    answer = os.getenv("FORGE_INSTALL_DEPENDENCIES")
+    if not answer:
+        answer = input("Would you like to install it now? (y/n) ").strip().lower()
+    if answer in ("y", "yes", "yeet", "yessir", "si"):
```
Comment (banool) [inline testsuite/forge.py:67]: nice lol

File: testsuite/forge.py [hunk 11]
```diff
@@ -0,0 +1,815 @@
...
+        shell = LocalShell(True)
+        shell.run(["pip3", "install", dependency], stream_output=True).unwrap()
+    else:
+        print(f"Please install click (pip install {dependency}) and try again")
```
Comment (banool) [inline testsuite/forge.py:71]: print to stderr?

File: testsuite/forge.py [hunk 12]
```diff
@@ -0,0 +1,815 @@
...
+
+def get_utc_timestamp(dt: datetime) -> str:
+    return dt.strftime("%Y-%m-%dT%H:%M:%S.000Z")
+
+
+@click.group()
+def main() -> None:
+    # Check that the current directory is the root of the repository.
+    if not os.path.exists('.git'):
+        print('This script must be run from the root of the repository.')
```
Comment (banool) [inline testsuite/forge.py:100]: Same comment here with printing to stderr. Same thing for anywhere else we print and then exit.

**Review by rustielin — APPROVED** [review]
LGTM :shipit:  gogogogoggo!!! 

Just remember to remove the push on `chaotic` bit, and to resolve the merge conflict.

**Pull request merged**